File: backend/domain/chain.py
"""
Chain domain — AI compute chain types, enums, and constants.
Pure logic, no I/O, no framework imports.
"""

from enum import Enum
from datetime import datetime
from dataclasses import dataclass
from typing import Optional, List


class ChainID(str, Enum):
    AI_COMPUTE = "ai-compute"


class RelationshipStatus(str, Enum):
    VERIFIED = "verified"
    CURATED = "curated"
    CANDIDATE = "candidate"


class EvidenceTier(str, Enum):
    PRIMARY_FACT = "primary_fact"
    SECONDARY_FACT = "secondary_fact"
    BROKER_OPINION = "broker_opinion"
    NEWS_EVENT = "news_event"
    MANUAL_SEED = "manual_seed"
    LLM_CANDIDATE = "llm_candidate"


class ConfidenceBand(str, Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class StrengthBand(str, Enum):
    STRONG = "strong"
    MODERATE = "moderate"
    WEAK = "weak"


class ExposureBand(str, Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"
    MINIMAL = "minimal"


class StalenessStatus(str, Enum):
    FRESH = "fresh"
    STALE = "stale"
    EXPIRED = "expired"


class ValidationStatus(str, Enum):
    NOT_YET_OBSERVABLE = "not_yet_observable"
    VALIDATED = "validated"
    PARTIALLY_VALIDATED = "partially_validated"
    CONTRADICTED = "contradicted"
    INSUFFICIENT_DATA = "insufficient_data"


class ImpactDirection(str, Enum):
    POSITIVE = "positive"
    NEGATIVE = "negative"
    MIXED = "mixed"
    UNKNOWN = "unknown"


class EventClassification(str, Enum):
    CAPEX_INCREASE = "capex_increase"
    CAPEX_CUT = "capex_cut"
    GPU_NEW_GENERATION = "gpu_new_generation"
    GPU_SUPPLY_CONSTRAINT = "gpu_supply_constraint"
    LIQUID_COOLING_ADOPTION = "liquid_cooling_adoption"
    OPTICAL_SPEED_UPGRADE = "optical_speed_upgrade"
    RAW_MATERIAL_PRICE_UP = "raw_material_price_up"
    RAW_MATERIAL_PRICE_DOWN = "raw_material_price_down"
    EXPORT_RESTRICTION = "export_restriction"
    DOMESTIC_SUBSTITUTION = "domestic_substitution"
    DEMAND_SLOWDOWN = "demand_slowdown"
    MANUAL_REQUIRED = "manual_required"
    UNKNOWN = "unknown"


@dataclass
class ChainNode:
    node_id: str
    chain_id: ChainID
    name: str
    level: str  # L1 or L2
    tier: str = "general"  # core, important, general
    description: Optional[str] = None
    upstream_node_ids: Optional[List[str]] = None
    downstream_node_ids: Optional[List[str]] = None
    key_technologies: Optional[List[str]] = None
    market_size: Optional[str] = None
    concentration: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


@dataclass
class CompanyMaster:
    company_id: str
    symbol: str
    exchange: str
    canonical_name: str
    aliases: List[str]
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


@dataclass
class CompanyNodeExposure:
    exposure_id: str
    company_id: str
    node_id: str
    chain_id: ChainID
    role: str
    exposure_band: ExposureBand
    business_scope: Optional[str] = None
    evidence_tier: EvidenceTier = EvidenceTier.MANUAL_SEED
    relationship_status: RelationshipStatus = RelationshipStatus.CURATED
    evidence_as_of: Optional[datetime] = None
    last_seen_date: Optional[datetime] = None
    staleness_status: StalenessStatus = StalenessStatus.FRESH
    review_required_after: Optional[datetime] = None
    effective_from: Optional[datetime] = None
    effective_to: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


@dataclass
class ChainEdge:
    edge_id: str
    chain_id: ChainID
    source_node_id: str
    target_node_id: str
    relationship_type: str
    strength_band: StrengthBand
    confidence_band: ConfidenceBand
    direction: str = "downstream"  # upstream, downstream, bidirectional
    layer: str = "general"  # core, important, general
    value_chain_position: Optional[str] = None
    evidence_tier: EvidenceTier = EvidenceTier.MANUAL_SEED
    relationship_status: RelationshipStatus = RelationshipStatus.CURATED
    description: Optional[str] = None
    business_logic: Optional[str] = None
    key_dependencies: Optional[List[str]] = None
    evidence_as_of: Optional[datetime] = None
    last_seen_date: Optional[datetime] = None
    staleness_status: StalenessStatus = StalenessStatus.FRESH
    review_required_after: Optional[datetime] = None
    effective_from: Optional[datetime] = None
    effective_to: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


@dataclass
class EventRule:
    rule_id: str
    chain_id: ChainID
    event_type: EventClassification
    affected_node_ids: List[str]
    impact_direction: ImpactDirection
    impact_level: str
    strength_band: StrengthBand
    confidence_band: ConfidenceBand
    time_lag_days: Optional[int] = None
    condition: Optional[str] = None
    exclusion: Optional[str] = None
    risk_factors: Optional[List[str]] = None
    rationale: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


@dataclass
class Evidence:
    evidence_id: str
    chain_id: ChainID
    source_type: EvidenceTier
    source_reference: str
    content: str
    confidence_band: ConfidenceBand
    effective_date: Optional[datetime] = None
    expiration_date: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


@dataclass
class EventAnalysisSnapshot:
    snapshot_id: str
    event_text: str
    normalized_event_type: EventClassification
    event_date: datetime
    classification_confidence_band: ConfidenceBand
    rules_version: str
    chain_data_version: str
    financial_data_as_of: datetime
    result_json: dict
    created_at: Optional[datetime] = None

File: frontend/src/types/chain.ts
/**
 * AI计算链相关类型定义
 */

// 链节点
export interface ChainNode {
  node_id: string;
  chain_id: string;
  name: string;
  level: string; // L1 or L2
  tier?: string; // core, important, general
  description?: string;
  upstream_node_ids?: string[];
  downstream_node_ids?: string[];
  key_technologies?: string[];
  market_size?: string;
  concentration?: string;
  representative_companies?: CompanyWithExposure[];
}

// 公司基本信息
export interface Company {
  company_id: string;
  symbol: string;
  exchange: string;
  canonical_name: string;
  aliases: string[];
}

// 带曝光信息的公司
export interface CompanyWithExposure extends Company {
  role?: string;
  exposure_band?: string;
  relationship_status?: string;
}

// 曝光记录
export interface CompanyNodeExposure {
  exposure_id: string;
  company_id: string;
  node_id: string;
  chain_id: string;
  role: string;
  exposure_band: string;
  business_scope?: string;
  evidence_tier: string;
  relationship_status: string;
  evidence_as_of?: string;
  last_seen_date?: string;
  staleness_status: string;
  review_required_after?: string;
  effective_from?: string;
  effective_to?: string;
  node?: ChainNode;
}

// 链边
export interface ChainEdge {
  edge_id: string;
  chain_id: string;
  source_node_id: string;
  target_node_id: string;
  relationship_type: string;
  direction?: string; // upstream, downstream, bidirectional
  layer?: string; // core, important, general
  value_chain_position?: string;
  strength_band: string;
  confidence_band: string;
  evidence_tier: string;
  relationship_status: string;
  description?: string;
  business_logic?: string;
  key_dependencies?: string[];
  evidence_as_of?: string;
  last_seen_date?: string;
  staleness_status: string;
  review_required_after?: string;
  effective_from?: string;
  effective_to?: string;
}

// 链概览
export interface ChainOverview {
  chain_id: string;
  nodes: ChainNode[];
  edges: ChainEdge[];
  companies: Company[];
  metadata: {
    node_count: number;
    edge_count: number;
    company_count: number;
    loaded_at: string;
  };
}

// 公司链定位
export interface CompanyChainPosition {
  status: 'mapped' | 'unmapped';
  message?: string;
  company_id?: string;
  company?: Company;
  exposures: CompanyNodeExposure[];
  node_count?: number;
  chain_id?: string;
  chain_context?: ChainContext;
}

// 链路上下文
export interface ChainContext {
  upstream_nodes: ChainContextNode[];
  downstream_nodes: ChainContextNode[];
  related_edges: ChainContextEdge[];
}

// 链路节点（带公司信息）
export interface ChainContextNode {
  node_id: string;
  name: string;
  level: string;
  tier: string;
  description: string;
  companies: CompanyWithExposure[];
}

// 链路边
export interface ChainContextEdge {
  edge_id: string;
  source_node_id: string;
  target_node_id: string;
  relationship_type: string;
  direction: string;
  layer: string;
  value_chain_position: string;
  description: string;
  business_logic: string;
}

// 事件分类结果
export interface EventClassification {
  event_type: string;
  confidence_band: string;
  matched_keyword?: string;
  method: string;
}

// 影响路径
export interface ImpactPath {
  rule_id: string;
  event_type: string;
  impact_direction: string;
  impact_level: string;
  strength_band: string;
  confidence_band: string;
  affected_node_ids: string[];
  time_lag_days?: number;
  rationale?: string;
}

// 事件影响分析结果
export interface EventImpactAnalysis {
  event_type: string;
  affected_nodes: ChainNode[];
  impact_paths: ImpactPath[];
  affected_node_count: number;
  rule_count: number;
}

// 基本面验证结果
export interface FundamentalValidation {
  status: string;
  symbol: string;
  signals: Record<string, any>;
  validated_at?: string;
  message?: string;
}

// 事件分析快照
export interface EventAnalysisSnapshot {
  snapshot_id: string;
  event_text: string;
  normalized_event_type: string;
  event_date: string;
  classification_confidence_band: string;
  rules_version: string;
  chain_data_version: string;
  financial_data_as_of: string;
  result_json: Record<string, any>;
  created_at: string;
}

// 事件分析请求
export interface AnalyzeEventImpactRequest {
  event_text: string;
  event_date?: string;
}

// 事件分析响应
export interface AnalyzeEventImpactResponse {
  classification: EventClassification;
  impact: EventImpactAnalysis | null;
  fundamental_validation?: Record<string, FundamentalValidation>;
  snapshot?: {
    snapshot_id: string;
    created_at: string;
    status: string;
  };
  message?: string;
}

// 事件规则
export interface EventRule {
  rule_id: string;
  chain_id: string;
  event_type: string;
  affected_node_ids: string[];
  impact_direction: string;
  impact_level: string;
  strength_band: string;
  confidence_band: string;
  time_lag_days?: number;
  condition?: string;
  exclusion?: string;
  risk_factors?: string[];
  rationale?: string;
}

// 证据记录
export interface Evidence {
  evidence_id: string;
  chain_id: string;
  source_type: string;
  source_reference: string;
  content: string;
  confidence_band: string;
  effective_date?: string;
  expiration_date?: string;
}